### ISSUE TITLE
Migrate JDBC tracing info from IBM Docs

### ISSUE THREAD

**Issue by dmuelle** [OpenLiberty/docs#5301]
from the [Liberty docs issue](https://github.ibm.com/websphere/liberty-docs/issues/2804) opened by @KyleAure 

Our support staff would like to reference customers to this page by default for all salesforce cases opened against the JDBC feature to show users how to enable JDBC tracing. 
However, the layout of the original document is confusing, the procedure is fragmented, and overall customers become more confused when reading this document.

Having this document migrated to the `openliberty.io` site would give the support and development teams the ability to update this document and make improvements based on customer feedback. 

To start with, migrate the original document as it is today.
Once it is migrated, the support and development team will suggest updates to make the document more accurate and user-friendly. 
Once finalized we can release it for customer use. 

 https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-enabling-jdbc-tracing
migrate to the following page, or create a new child page
https://www.openliberty.io/docs/21.0.0.3/log-trace-configuration.html

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle 

For review the migrated doc :
https://draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

Regards,
Ramkumar

**Comment by KyleAure** [comment]
@ramkumar-k-9286 and @dmuelle 
I made a PR to make some initial changes to this doc under #5423

**Comment by dmuelle** [comment]
Thank you @KyleAure! @ramkumar-k-9286 to avoid future branch conflicts, you might manually copy Kyle's changes into your branch and commit them from there. Otherwise, when you commit the draft to staging from your working branch, it might not have all the changes.

**Comment by KyleAure** [comment]
@ramkumar-k-9286 Would you mind holding off releasing this till next week (not sure of your release schedule)?  I want to have our L2 support also review this documentation before we put it into our doc. I wasn't able to schedule a meeting with them until monday.

**Comment by dmuelle** [comment]
Hi @KyleAure - Ram is on vacation today but the work on this draft is ongoing. We need to do some revisions to bring into line with Open Liberty content standards- in addition to any needed edits to technical information.  We are working on reorganizing the information to make it more straightforward and easier to follow, with less redundancy.

Right now we are targeting the revised doc to release in 22.0.0.6 on 6/7/22 so we have some time work on it. If L2 provides any feedback on Monday we can work it in to our ongoing edits and then share the revised draft for final review and make any further edits that you recommend.

**Comment by dmuelle** [comment]
@KyleAure - we have a few questions to help clarify and organize some of the info in this doc. If it's helpful, please share these in your meeting with L2. I'm also happy to join the meeting if that would be helpful. Let me know you you think.

The doc is hard to follow because it provides repetitive examples without adequately explaining what the differences between them are, what syntax patterns the configuration follows, or when you might need one option vs the other, or both. We need better baseline explanations for each case- java.util, custom, vs supplemental. 

## Use java.util.logging
- To use the Java logging API for JDBC drivers that support the java.util.logging, you must specify: 
**com.ibm.ws.logging.trace.specification=*=audit:<_driver name_>=<_trace log level_>**
in the bootstrap.properties file. Is that correct?
  - No, for jdbc drivers that use java.util.logging users will need to specify the *package name* associated with the classes within the driver.  We should be providing a list of the drivers that use Java logging, and their associated package names.
- can you also set this in the server.xml by using
 **<logging traceSpecification="*=audit:<_driver name_>=<_trace log level_>/>** ?
  - Yes, but instead you would want to use **<logging traceSpecification="*=audit:<_package name_>=<_trace log level_>/>** 
- Do you need the vendor log writer at all if you just want to use the Java logger?
  -  No, since Liberty has default config for java.util.logging the user can't provide a separate log writer for JDBC drivers. 
- For Oracle, you have to also set the `oracle.jdbc.Trace` system property to true. Is Oracle the only vendor that needs a system property?
  - Yes, Oracle is the only driver that requires an additional system property.  Most other drivers that require additional configuration will do so via Datasource configuration.  In addition, oracle has two different drivers, one is the normal driver, and one is the debug driver.  For trace to be produced uses have to use the debug driver.  
  
## Use custom trace settings

- Does this section describe how to enable all JDBC tracing for drivers that don't support the Java logging API? Or does it describe how to set custom trace properties for any JDBC driver that has them? It's hard to tell if this section is an alternative to using the java logger or something that would supplement default driver settings.
  - This section is for drivers that both support java.util.logging and require additional properties to be configured on their datasource in order for the driver to actually produce trace. 
- What constitutes a "custom" trace setting in a Liberty context and how do know if you need to specify one?
  - A user would only know if they needed a custom trace setting if they read the documentation for their driver carefully.  Or read this documentation.  Which is why we want to document it. 
- Syntax- for any custom trace settings, you specify them in the server.xml in a `dataSource` element in a `properties.$vendor.etc` sub-element for your driver. But in order for them to work, you also have to add the specified log writer to your `bootstrap.properties` file. Is that correct?
  - Correct, if a user only provides the custom trace settings, then trace will be produced by the driver, but the log writer for Open Liberty won't write them to the trace file. 

## Use supplemental JDBC tracing
- I'm assuming in a Liberty context, "the application server" = "Liberty" and we'll make that edit in the doc unless that's incorrect. 
  - Correct 
- what does supplemental trace enable? why would you want it?
  - The supplementation trace should only be used for JDBC drivers that DO NOT use the java.util.logging library to produce trace.  Typically, this means the driver is using a basic PrintWriter to directly log to a file instead of using java.util.logging, or they are using a thrid-party logging library like log4j or slf4j. Without being able to leverage the java.util.logging library to collect trace from the driver, we have `logwriters` setup to intercept method calls at the boundry of the JDBC specification and provide our own trace. For example we will log the enter/exit of the getConnection() method. 

>  The application server automatically determines whether to enable supplemental JDBC tracing, based on the JDBC driver used.
  - This statement is misleading.  It sounds like with no user action we will take over and enable JDBC tracing by default. We do not. If a user tries to use a logwriter for a driver that supports java.util.logging then we will typically DISABLE the log writer because it will produce inferior trace to that provided by the JDBC driver itself. 
 
> To override automatic determination on enabling supplemental JDBC tracing, set the supplementalJDBCTrace data source property to true or false.
  - The supplementalJDBCTrace attribute can override that decision, but also doesn't always work.  Like in the case of the Oracle Driver. I think it would be better to remove mention of this property as it is leftover from TWAS and currently doesn't work as specified here. 

- why would you need to override automatic determination? 
  - You should not need to. 
- what is the default for supplementalJDBCTrace? 
  -  The default is null (unset). 
- do you only set  supplementalJDBCTrace  when supplemental trace is not enabled by default and you need supplemental tracing? Do you only set it to false when supplemental trace is enabled by default and you don't want it?
  - The original use was to allow you to enable both supplemental and java.util.logging trace at the same time.  But really we should direct customers to use one or the other. 
- if you are using a driver for which supplemental tracing is enabled by default, do you still need to add the log writer to your bootstrap.properties file?
  - Supplemental tracing is never enabled by default. Users will never need to add a log writer to bootstrap.properties specifically for JDBC tracing since bootstrap.properties is specific to server config. 

Thanks!

**Comment by KyleAure** [comment]
@dmuelle @ramkumar-k-9286 I updated the previous comment with the answers to the questions you asked.

**Comment by KyleAure** [comment]
FYI I opened PR #5469 with some suggested changes I collected from our L2 support team.  I also went through and tested most of these configurations to ensure they work.

**Comment by dmuelle** [comment]
Thanks for this info @KyleAure - we have incorporated the changes from your PR into draft by copying them into Rams working branch and committing from there instead of from your PR. This is onl;y because of the way the Open Liberty publishing flow is set up- since `draft` is not strictly in sync with `staging` and `vNext`, it is much easier to work on a doc where all the changes for a particular issue/release come from a single working branch.

We will continue working on the doc and will send back to you for review. We are hoping to publish with 22.0.0.6, but might push to a future release if needed. Thanks!

**Comment by KyleAure** [comment]
Perfect that's what I expected.  Thanks again for the hard work and getting this document into shape!

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Incorporated changes suggested in Kyle's PR. 
Draft link: https://draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

Regards,
Ramkumar

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

Reworked the draft based on your comments yesterday.

Draft link: https://draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

Regards,
Ramkumar

**Comment by dmuelle** [comment]
Looks good- a few suggested edits.

## short description
I think this could be more concise and less repetitive

JDBC driver tracing for Open Liberty enables tracing for your third-party JDBC driver for debugging purposes, or to submit a ticket. To collect JDBC driver trace, you need to enable tracing for your JDBC driver in Open Liberty, configure the logging level, and optionally provide driver-specific configurations when applicable.

--->

You can enable tracing for your third-party JDBC driver in Open Liberty for debugging and support purposes. To collect JDBC driver trace, you must specify your JDBC driver, set the logging level, and provide any required driver-specific configurations.

## Understanding your JDBC driver
Recommend not using "JUL" abbreviation- we've never used it in Liberty docs, IBVM terminology doesn't recognize it,  and oracle doesn't seem to use it either.

A third-party JDBC driver uses the native Java logging library, or some other logging library such as SLF4J or Log4J. If your JDBC driver uses the Java utility logging(JUL) library, you can enable logging for your JDBC driver by providing a trace specification corresponding to the package name that is used by the driver. If your JDBC driver uses some other logging library, then Open Liberty provides a logwriter that generates logs based on the JDBC specification methods like getConnection, executeQuery etc.

-->

A third-party JDBC driver uses either the `java.util.logging` Java logging library or some other logging library, such as SLF4J or Log4J. If your JDBC driver uses the `java.util.logging` library, you can enable trace by providing a trace specification that corresponds to the package name for the driver. If your JDBC driver uses some other logging library, Open Liberty provides a log writer that generates logs based on JDBC specification methods such as `getConnection` and `executeQuery`.


The following table lists the most common JDBC drivers, their support for JUL, and the recommended trace specification (package name or logwriter).
--->
The following table lists the most common JDBC drivers, whether they support the `java.util.logging` library, and the package name or log writer to use in the trace specification.


The following sub sections contain JDBC driver specific configuration.
--->
The following examples demonstrate driver-specific trace configuration for commonly used JDBC drivers.

**Comment by ramkumar-k-9286** [comment]
Hi David @dmuelle

suggested corrections made to the JDBC doc.

Draft link: https://draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

Regards,
Ramkumar

**Comment by KyleAure** [comment]
@ramkumar-k-9286 WOW this looks great.  Here are just a few suggestions (that I forgot).

DB2 JCC:
Add an additional table to specify the other values for traceLevel, or link to this document: 
https://www.ibm.com/support/pages/collecting-data-tracing-ibm-data-server-driver-jdbc-and-sqlj
(see table 1)

Informix using DB2 JCC:
Should have that same table for traceLevel, or link to this document: https://www.ibm.com/support/pages/collecting-data-tracing-ibm-data-server-driver-jdbc-and-sqlj

Oracle:
Add the following text before the configurations:

```
Oracle provides two different drivers one for production and one for debug purposes. The production driver does not produce trace, so you will need to download and swap out your production driver, for the debug driver. The debug driver will have `_g` in the driver name. (i.e. ojdbc8.jar will be ojdbc8_g.jar)
```

Thank you,
Kyle Jon Aure

**Comment by dmuelle** [comment]
@KyleAure @ramkumar-k-9286 - this is shaping up! I think we could make it more clear what is required to _enable_ trace vs. what you can do to _configure_ it. As we have it, these two user goals (enabling and configuring) are kind of muddled together. 

Ideally, we would 
1. explain what users need to do to enable trace, with representative examples for JUL and non-JUL scenarios
2. explain how to configure custom settings, with a representative example of setting vendor property (like the DB2 JCC example)
3. provide the driver specific config examples, which should include any required extra config in the section for each driver- such as the oracle system property and info about the debug tracer and the DB2 trace level stuff.

We are close to having this now but the custom settings and enablement examples are mixed together so that its tricky to tell what is enablement and what is custom config. The explanation sections (1,2) don't need exhaustive examples - just one or two to make the explanation clear. 


Right now in the "Understanding Open Liberty tracing section" we have this statement
```
For driver-specific custom trace facilities, you can use one of the following two options.

  - Use the Java built-in logging mechanism, java.util.logging.

  - Configure a custom trace setting as a vendor property.
```

But I don't think these are two mutually exclusive options that the user will choose between. Maybe instead something like


> How you enable trace for your JDBC driver depends on whether the driver supports the ` java.util.logging` logging library. In either case, you must identify your driver in the trace specification. Drivers that support the ` java.util.logging` library are specified by the package name that the driver uses. Drivers that do not support the ` java.util.logging` library are specified either by the log writer name for the driver or by the Open Liberty generic log writer name. 

> When trace is enabled for your driver, you can configure custom trace settings as vendor properties.

 

@KyleAure - a few more questions :) 

Does the table of 3rd party drivers provide value when we already provide enable/config examples for all the drivers in the table? Couldn't users just get the example they need from the examples and not worry about whether or not their driver supports JUL or if they're specifying a package name or log writer name? If theres nothing in particular a user can _do_ with the info in the table (other than just know) could we leave it out altogether and just provide the examples?


I'm trying to understand what the "Recommended trace configuration for support" section is showing. I _think_ its showing trace enablement and configuration but leaving out the driver details- which presumably the user would add for their driver, based on the info we've provided elsewhere- is that correct?

Thanks again

**Comment by KyleAure** [comment]
@dmuelle 
> Does the table of 3rd party drivers provide value when we already provide enable/config examples for all the drivers in the table? 

I agree that the table can be removed in lieu of the individual sections.
I think that we should document which drivers use Java Util Logging, and which drivers just use our integrated logwriters.
Does it affect what the user has to configure? No
However, letting the user know that JUL will produce trace from the driver itself vs. our logwriters which will perform logging for the driver is an important distinction if they are enabling this trace for their own debugging purposes. 

> I'm trying to understand what the "Recommended trace configuration for support" section is showing.
I want this document to be robust enough for our support staff to be able to send this to customers when we ask for a trace. 

Our support software has certain limitations as far as to file sizes so we want to make sure customers don't produce logs that are too large for us to parse. 
In addition, we want to make sure we don't produce so many logs as to fill up the user's system storage. 
This leads into the next statement

> I think we could make it more clear what is required to enable trace vs. what you can do to configure it.

I agree, so far what we have is mostly just the enablement part without much description of what can be configured by the user. I think we should make the changes suggested thus far and create a new draft and then start parsing out what's the is configurable vs. required.

**Comment by KyleAure** [comment]
@ramkumar-k-9286 Here is some additional feedback from our L2 support team

1. Can we include that after altering the bootstrap.properties file, you must restart the server for the changes to take effect (this used to be the case and I believe still is).
2. Maybe do a bit of explanation regarding Oracles `-Doracle.jdbc.Trace=true` option since it is the only trace in which you need to set an additional property in another file or set it programatically. It feels like this could use some further information or at least a link to Oracle’s documentation explaining it.
    - The explanation would be, "The debug driver that produces trace is conservative with the trace it produces.  To provide the highest quality of service we recommend setting `-Doracle.jdbc.Trace=true` to ensure that all trace is produced."
3. The way it’s phrased right now it seems a little ambiguous if/when you have to use the bootstrap.properties file since it says “You can also add the logwriter name to the trace specification in the bootstrap.properties file.” I think this should be a bit more explicit in saying that it can be used as an alternative (although I seem to remember there being certain scenarios where it has to be used)
    - Perhaps the best way to resolve this would be to have all config examples only use server.xml, and add a note somewhere to state that the same configuration could be done using bootstrapping.properties instead? This would make suggestion (1) obsolute though. 

Let me know if this is getting too complicated or not.  The topic itself is a bit complicated which is why we want to document it correctly so customers have an easier time gathering trace.

**Comment by dmuelle** [comment]
> Perhaps the best way to resolve this would be to have all config examples only use server.xml, and add a note somewhere to state that the same configuration could be done using bootstrapping.properties instead? This would make suggestion (1) obsolute though. 

I agree that it would be more clear if we stuck to just the server.xml examples- it's almost always better to just tell someone one clear path to a goal rather than give multiple equivalent options that they then have to choose between. We could demonstrate how to configure in the bootstrap.properties as an alternative without giving exhaustive examples for every DB- and also mention w/ that example that you have to restart the server for it to take effect.

**Comment by dmuelle** [comment]
@ramkumar-k-9286 

feedback for recent changes

## Understanding Open Liberty tracing
Drivers that do not support the java.util.logging or uses some other logging library, Open Liberty provides a log writer that generates logs based on JDBC specification methods such as getConnection and executeQuery.
--->
Drivers that do not support the` java.util.logging library`, are specified by the log writer name for the JDBC driver. Open Liberty provides log writers for common database vendors that generate logs based on JDBC specification methods such as `getConnection` and `executeQuery`. Open Liberty also provides a generic log writer for drivers that do not support  the `java.util.logging` library and for which no vendor-specific log writer exists.

For more information on Open Liberty tracing, see https://openliberty.io/docs/latest/log-trace-configuration.html[Log and trace configuration].
--->
For more information on Open Liberty tracing, see xref:log-trace-configuration.adoc[Log and trace configuration].
## Enable trace specification
Enable trace specification
--->
Enable trace for your JDBC driver

add to the beginning of this section- 

> The following examples demonstrate how to enable trace for JDBC drivers. For vendor-specific examples that include the necessary package or log writer names for common JDBC drivers, see <<#anchor,Trace configuration examples for common JDBC drivers>>.

For JDBC drivers that support the java.util.logging, you will need to specify the package name that the driver uses. You can set this in the server.xml file.
--->
For JDBC drivers that support the java.util.logging, specify the package name that the driver uses, as shown in the following `server.xml` file example for the Microsoft SQL JDBC driver.


For JDBC drivers that do not support the java.util.logging or uses some other logging library, Open Liberty provides a log writer that generates logs based on JDBC specification methods. You can set this in the server.xml file.
--->
For JDBC drivers that do not support the `java.util.logging`  library, specify either the name of a vendor-specific log writer or the generic Open Liberty log writer. The following `server.xml` file example shows the trace specification for the Derby JDBC driver, which does not support the `java.util.logging` library.


Alternatively, you can also add the logwriter name to the trace specification in the bootstrap.properties file.
-->
Logging is initialized before the` server.xml` file is processed so configuring logging through the `server.xml` file can result in early log entries that use a different log configuration from later ones. To avoid this problem, you can configure the trace specification in your `bootstrap.properties` file instead. The following `bootstrap.properties` file example shows the trace specification for the Microsoft SQL Server DataDirect driver, which does not support the `java.util.logging` library.

{EXAMPLE}
 Drivers that support the `java.util.logging` library can also be configured in the `bootstrap.properties` file by specifying the package name for the driver in stead of a log writer name.


## Configure custom settings
If the JDBC driver that you are using has custom trace settings
--->
If your JDBC driver has custom trace settings

Remove the bootstrap example under this section.

## Trace configuration examples for common JDBC drivers

As an example of how to revise the example introductions in the vendor-specific section:

The DB2 JCC driver supports the java.util.logging library. The driver uses the com.ibm.db2 in the trace specification.
--->
The DB2 JCC driver supports the `java.util.logging` library. The driver uses the `com.ibm.db2` package name in the trace specification. The following `server.xml` file example shows the configuration to enable trace for the DB2 JCC driver and specify a custom property.


When you rewrite these, watch out for

does not supports
--->
does not support

Check the whole topic for typos, I see a few here and there.

**Comment by ramkumar-k-9286** [comment]
Hi Kyle @KyleAure 

The draft document ( based on suggestions from you and David):

link: https://docs-draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

Could you suggest an explanation for the `library` config in the Oracle example?

Regards,
Ramkumar

CC @dmuelle

**Comment by KyleAure** [comment]
@ramkumar-k-9286 Thanks for sending this along. 

## Oracle - Library Explanation

### Current

Oracle provides two different drivers, one for production and another for debugging purposes. The production driver does not produce a trace, so you need to download and replace your production driver with the debugging driver. The debugging driver has _g in the driver name. For example, ojdbc8.jar is ojdbc8_g.jar.

```xml
<logging traceSpecification="*=info:RRA=all:oracle=all" />
<library id="oracleDebug">
    <file name="${path.to.oracle.dir}/ojdbcX_g.jar"/>
</library>
```

### Suggested

Oracle provides two different drivers, one for production and another for debugging purposes. The production driver does not produce any trace, so you need to download and replace your production driver with the debugging driver. The debugging driver has _g in the driver name. For example, ojdbc8.jar is ojdbc8_g.jar.  The debugging driver will need to be configured as the jdbcDriver element in server.xml.

```xml
<logging traceSpecification="*=info:RRA=all:oracle=all" />
<jdbcDriver id="oracleDriver">
  <library id="oracleDebug">
      <file name="${path.to.oracle.dir}/ojdbcX_g.jar"/>
  </library>
</jdbcDriver
```

## Oracle - JVM Option

### Current

The trace produced debugging driver is limited. You can add the trace configuration to the jvm.options file to ensure that optimum trace is produced.

```txt
-Doracle.jdbc.Trace=true
```

### Suggested

The trace produced by the debugging driver is limited. You can add the following option to the jvm.options file to ensure that all trace is produced.

```txt
-Doracle.jdbc.Trace=true
```


Otherwise, this looks really great!  
Thanks for all the hard work.

**Comment by ramkumar-k-9286** [comment]
Hi Kyle @KyleAure

The suggested changes have been incorporated.

link: https://docs-draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html


Regards,
Ramkumar

CC @dmuelle

**Comment by dmuelle** [comment]
## Peer review

### Understanding Open Liberty tracing

> When trace is enabled for your driver, you can **configure custom trace settings as vendor properties**.
Make this a link to the "Configure custom settings" section.

### Oracle example

The debugging driver needs to be configured as the jdbcDriver element in server.xml.
--->
Configure the debugging driver by specifying the `library subelement within the  `jdbcDriver` element. The library defines the path to the debugging driver JAR file.
___
The following server.xml file example shows you the configuration to enable trace for the Oracle driver.
--->
The following `server.xml` file example shows you how to configure the Oracle debugging driver and  enable trace.
___
In the code example, we only use ${this syntax} for predefined server variables that can be directly entered in config and automatically replaced at run time, eg `${server.config.dir}`

${path.to.oracle.dir}/ojdbcX_g.jar
--->
path_to_oracle_driver/ojdbcX_g.jar

also `</jdbcDriver` is missing the closing brace `>`
___
The trace produced by the debugging driver is limited
--->
The trace that is produced by the debugging driver is limited.

**Comment by dmuelle** [comment]
Also- add a link to this doc from the [JDBC driver library configuration section of the Relational database connections with JDBC doc](https://draft-openlibertyio.mybluemix.net/docs/latest/relational-database-connections-JDBC.html#_jdbc_driver_library_configuration).

At the end of the section, something like, "For information about configuring trace for your JDBC driver, see **JDBC tracing**

**Comment by ramkumar-k-9286** [comment]
Peer Review comments have been incorporated.

Added a link to JDBC tracing doc from Relational Database Connections JDBC doc.

Links :

https://docs-draft-openlibertyio.mybluemix.net/docs/latest/jdbc-tracing.html

https://docs-draft-openlibertyio.mybluemix.net/docs/latest/relational-database-connections-JDBC.html#_jdbc_driver_library_configuration

**Comment by dmuelle** [comment]
-  link to JDBC tracing doc from Relational Database Connections JDBC doc is not working

Configure the debugging driver by specifying the library subelement within the `jdbcDriver` element. The library defines the path to the debugging driver JAR file.
---->
Configure the debugging driver by specifying the `library` subelement within the `jdbcDriver` element. The `library` subelement defines the path to the debugging driver JAR file.

**Comment by dmuelle** [comment]
content is on vNext and will publish with 22006, closing this issue.
